Userland/Libraries/LibGfx/PathPainter.cpp

```diff
@@ -0,0 +1,215 @@
+    }
+}
+
+void PathPainter::begin_path(StrokeKind stroke_kind, FillKind fill_kind, float thickness)
+{
+    m_stroke_kind = stroke_kind;
+    m_fill_kind = fill_kind;
+    m_thickness = thickness;
+}
+
+void PathPainter::move_to(Point<float> p)
+{
+    check_end();
+    m_position = p;
+}
+
+void PathPainter::line_to(
+    Point<float> p2)
+{
+    check_begin();
+    edge_to(p2, true);
+    m_position = p2;
+}
+
+void PathPainter::quadratic_bezier_curve_to(
+    Point<float> p2,
+    Point<float> p3)
+{
+    check_begin();
+    flatten_quadratic_bezier(m_position, p2, p3);
+    m_position = p3;
+}
+
+void PathPainter::cubic_bezier_curve_to(
+    Point<float> p2,
+    Point<float> p3,
+    Point<float> p4)
+{
+    check_begin();
+    flatten_cubic_bezier(m_position, p2, p3, p4);
+    m_position = p4;
+}
+
+void PathPainter::elliptical_arc_to(FloatPoint p2, FloatPoint center, FloatPoint radii, float x_axis_rotation, float theta_1, float theta_delta)
+{
+    check_begin();
+    Painter::for_each_line_segment_on_elliptical_arc(
+        m_position,
+        p2,
+        center,
+        radii,
+        x_axis_rotation,
+        theta_1,
+        theta_delta,
+        [this](FloatPoint /*from*/, FloatPoint to) {
+            edge_to(to, true);
+        });
+}
+
+void PathPainter::end()
+{
+    if (m_fill_kind == FillKind::Filled)
+        m_fill_painter.end();
+    switch (m_stroke_kind) {
+    case StrokeKind::OpenStroke:
+    case StrokeKind::ClosedStroke:
+        m_stroke_painter.end();
+        break;
+    case StrokeKind::NoStroke:
+        break;
+    }
+    m_in_path = false;
+}
+
+void PathPainter::end_path(Rasterizer::Paint const& stroke_paint, Rasterizer::Paint const& fill_paint)
+{
+    check_end();
+    switch (m_stroke_kind) {
+    case StrokeKind::OpenStroke:
+    case StrokeKind::ClosedStroke:
+        m_stroke_painter.end_path(stroke_paint);
+        break;
+    case StrokeKind::NoStroke:
+        break;
+    }
+    if (m_fill_kind == FillKind::Filled)
+        m_fill_painter.end_path(fill_paint);
+}
+
+void PathPainter::flatten_cubic_bezier(
+    Point<float> p1,
+    Point<float> p2,
+    Point<float> p3,
+    Point<float> p4,
+    int level,
+    bool corner)
+{
+    auto good_enough = [&] {
+        constexpr float tessTol = 0.25f;
+        auto d = p4 - p1;
+        if (AK::max(AK::abs(d.x()), AK::abs(d.y())) < tessTol)
+            return true;
+        auto a = AK::abs((p2 - p1).cross(d)) + AK::abs((p3 - p4).cross(d));
+        return a * a < tessTol * d.dot(d);
+    };
+    if (level > 10 || good_enough()) {
+        edge_to(p4, corner);
+    } else {
+        auto p12 = (p1 + p2) / 2;
+        auto p23 = (p2 + p3) / 2;
+        auto p34 = (p3 + p4) / 2;
+        auto p123 = (p12 + p23) / 2;
+        auto p234 = (p23 + p34) / 2;
+        auto p1234 = (p123 + p234) / 2;
+        flatten_cubic_bezier(p1, p12, p123, p1234, level + 1, true);
+        flatten_cubic_bezier(p1234, p234, p34, p4, level + 1, false);
+    }
+}
```

Comment:
Can you please clarify how this is or isn't related to https://github.com/memononen/nanovg/blob/7544c114e83db7cf67bd1c9e012349b70caacc2f/src/nanovg.c#L1286 . Same question for the rest of the code.

Reply:
Also the variable name `tessTol` is in camel case (and is needlessly abbreviated) which does not match serenity's style (though it does appear to be the same name that's used in NanoVG).

Reply:
> Can you please clarify how this is or isn't related to https://github.com/memononen/nanovg/blob/7544c114e83db7cf67bd1c9e012349b70caacc2f/src/nanovg.c#L1286 . Same question for the rest of the code.

Hi,

True, this code is tighly modeled from the corresponding function in nanosvg, the question is how exactly to change it, because it is basically algorithmic. The rest of the code is inspired by nanosvg but completely restructured. It is also strongly inspired, in architecture, by https://github.com/srwiley/rasterx

Reply:
Incidentally it also looks like the rasterizer that is still in `AnatialsingPainer`, and that I eventuall want to fully replace, is also strongly nanosvg inspired. in the end there are simply not that many ways to do this rasterization, similarly there are not too many ways to tesselate a bezier curve imho.

Reply:
I haven't personally looked at `AntiAliasingPainter` and - in general - I agree that there are only so many ways of doing things. However, to me this PR looks like a derived work from that other library, both in general code structure, variable names and how things are done specifically, almost as if the original C code was taken and "C++ified".

Reply:
Hmm, can you be more precise? In the case of the cubic bezier tesselation I did start from the original function. the rest I kinda disagree to be honest. as said, the rasterizer is simply an algorithmic decision, and the rest to be honest structurally is more inspired by the go rasterizer, where I didnt actually dive into the code very much.

Reply:
so would you be ok if I just rewwrite the bezier curve tesselation? I would still use the same approach though, i.e. use recursion to split the curve, because its simply the most straight forward way of doing it, and the recursion depth imho is neglegible.

Reply:
to make a point where I diverged a lot from nanosvg is the way strokes are generated: in nanosvg there is a lot of caching of points and reiterating, while I took the direction of rasterx and by breaking things up better I can avoid saving the original points and simultaneously tesselate the strokes and fills

Reply:
I am now going ahead and renaming and restructuring to make it more clear where "derivation" is purely in algorithmic nature. In the process I am also taking literal code from `Painter` regarding bezier curve splitting, still need to fix some bugs I introduces though